spylib/token.py

```diff
@@ -0,0 +1,428 @@
+                )
+            elif WRONG_OPERATION_NAME_ERROR_MESSAGE.format(operation_name) in errorlist:
+                raise ShopifyCallInvalidError(
+                    f'Store {self.store_name}: Operation name {operation_name}'
+                    'does not exist in the query.'
+                )
+            else:
+                raise ValueError(f'GraphQL query is incorrect:\n{errorlist}')
+
+        return jsondata['data']
+
+    def __new__(cls, *args, **kwargs):
+        """
+        Function to keep user from instantiating an instance of this class.
+        """
+        if cls is Token:
+            raise TypeError(f'Can\'t instantiate abstract class {cls.__name__} directly')
+        return object.__new__(cls)
+
+    def save_json(self):
+        json.dumps(
+            {
+                'store_name': self.store_name,
+                'scope': self.scope,
+                'access_token': self.access_token,
+                'access_token_invalid': self.access_token_invalid,
+                'client': self.client,
+                'oauth_url': self.oauth_url,
+                'api_url': self.api_url,
+                'api_version': self.api_version,
+                'rest_bucket_max': self.rest_bucket_max,
+                'rest_leak_rate': self.rest_leak_rate,
+                'graphql_bucket_max': self.graphql_bucket_max,
+                'graphql_leak_rate': self.graphql_leak_rate,
+            }
+        )
+
+
+class OfflineTokenABC(Token, ABC):
+    """
+    Offline tokens are used for long term access, and do not have a set expiry.
+    """
+
+    def __init__(self, **kwargs) -> None:
+        super().__init__(**kwargs)
+
+    async def reset_token(self, client_id: str, client_secret: str, code: str):
+
+        token: OfflineTokenResponse = OfflineTokenResponse(
+            **await self._get_token(client_id, client_secret, code)
+        )
+        self.access_token = token.access_token
+        self.scope = token.scope.split(',')
+
+
+@dataclass
+class OnlineTokenABC(Token):
+    """
+    Online tokens are used to implement applications that are authenticated with
+    a specific user's credentials. These extend on the original token, adding
+    in a user, its scope and an expiry time.
+    """
+
+    associated_user_scope: Optional[List[str]]
+    associated_user: Optional[AssociatedUser]
+    expires_at: Optional[datetime]
+
+    def __init__(
+        self,
+        associated_user_scope: Optional[List[str]] = [],
+        associated_user: Optional[AssociatedUser] = None,
+        expires_in: int = 0,
+        **kwargs,
+    ) -> None:
+        super().__init__(
+            **dict(
+                **kwargs,
+                **{
+                    'associated_user_scope': associated_user_scope,
+                    'associated_user': associated_user,
+                    'expires_at': datetime.now() + timedelta(days=0, seconds=expires_in),
+                },
+            )
+        )
```

Comment:
You shouldn't have to define `__init__`. That's one thing I really like with dataclasses and better in pydantic `BaseModel`